src/controllers/connectionManager.ts

```diff
@@ -462,6 +477,25 @@ export default class ConnectionManager {
 		);
 	}
 
+	public async showInstructionTextAsWarning(profile: IConnectionInfo, reconnectAction: IReconnectAction): Promise<void> {
+		const selection = await this.vscodeWrapper.showWarningMessageAdvanced(
+			LocalizedConstants.msgPromptSSLCertificateValidationFailed,
+			{ modal: false },
+			[
+				LocalizedConstants.enableTrustServerCertificate,
+				LocalizedConstants.readMore,
+				LocalizedConstants.cancel
+			]);
+		if (selection === LocalizedConstants.enableTrustServerCertificate) {
+			profile.encrypt = EncryptOptions.Mandatory;
+			profile.trustServerCertificate = true;
+			await reconnectAction(profile);
```

Comment:
Make sure you're being careful with these callbacks - if an error occurs in them (such as saveProfile throwing an error) then it'll silently fail. It'd be good to at the very least log the error in that case.

Reply:
saveProfile is done as part of 'reconnectAction' and the delegate takes care of throwing error.

Reply:
Yes, but nothing is handling the error. So if the callback/delegate throws then it will just silently fail.

Reply:
What do you mean? Reconnect action is awaited.

Reply:
But the calls to showInstructionTextAsWarning are not. 

Here's an example - you call showInstructionTextAsWarning here :

https://github.com/microsoft/vscode-mssql/pull/17484/files/050876a8d911f5456edd4cc4db39786a29bacc5d#diff-5eeace56def153787efa337fcd2f2e2771a202d038a1da07302cb41f6672a255R136

That call isn't awaited so it happily continues execution in that function. 

Meanwhile this function is running and eventually calls the reconnectAction. In this example let's say that the call to [saveProfile](https://github.com/microsoft/vscode-mssql/pull/17484/files/050876a8d911f5456edd4cc4db39786a29bacc5d#diff-5eeace56def153787efa337fcd2f2e2771a202d038a1da07302cb41f6672a255R141) throws an error. There's no catch in the delegate so it bubbles back up to this function here. This also doesn't have anything to catch it so it continues up the chain.

But because you didn't await (or have a .catch) on the call that error is never handled and lost forever.

Reply:
Yeah, I can't await either. I did test and noticed the error is thrown in console. I can add catch blocks to the calls, bt I don't think that's going to help either.

Reply:
Typically in that you'd either retry or display an error to the user so they at least know something went wrong. Displaying an error is probably easiest - although keep in mind that for this stuff one of the things you're trying to do is reconnect which is much more likely to fail (due to any number of issues from network problems to server being shut off....) and so that might impact the usefulness of what you're trying to do here. 

If you did consider a retry you could show an error dialog with the error and have a button for "Retry" that will allow them to continue retrying it as many times they wanted in case it seems like a transient error.

Reply:
Bug to track this: https://github.com/microsoft/vscode-mssql/issues/17501

Reply:
I've fixed this behavior for this PR, so we have it working, and on failure to save profile, we would get:

![image](https://user-images.githubusercontent.com/13396919/206331808-e4050d28-af42-4067-8014-8a90d58950d9.png)

I will fix the issue in main connection flow with the issue #17501